lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -1612,6 +1612,9 @@ bool StructuralEquivalenceContext::CheckKindSpecificEquivalence(
       if (!::IsStructurallyEquivalent(Name1, Name2) ||
           !::IsStructurallyEquivalent(*this, Record1, Record2))
         return false;
+      if (Record1->getQualifiedNameAsString() !=
```

Comment:
Qualified names may be different with any descendant of a `NamedDecl` can't they?
So perhaps this check could be more general at the start of `CheckKindSpecificEquivalence` ?

Reply:
Maybe yes the check can be for anything that has qualified name. But I was not sure if this is needed, we can even say that a `A::X` and `B::X` can be equivalent if the definition is equivalent. (But for `ASTImporter` not this is the good behavior.)

Reply:
Okay, then could you please add a FIXME/TODO about that we might want to do that check for NamedDecls?

Reply:
Or we could just have a test run on ci2 where we could examine the results of the change where all NamedDecls' qualified names are examined.

Reply:
The general check (in `CheckCommonEquivalence`) causes multiple test failures.

Reply:
Do the unit tests fail or the lit tests or both?

Reply:
In lit tests the ASTMerge/class-template test fails, this is caused by reordering of warnings and types in warnings.
The unit tests fail because there is difference in template argument names but these should be equivalent (template argument names should not be compared). This is not easy to fix (the qualified name contains the template parameter names).

Reply:
I think we could do the general check if we leave out all parameters:
```
 bool StructuralEquivalenceContext::CheckKindSpecificEquivalence(
     Decl *D1, Decl *D2) {
+
+  if (auto *ND1 = dyn_cast<NamedDecl>(D1)) {
+    auto *ND2 = dyn_cast<NamedDecl>(D2);
+    if (!ND2)
+      return false;
+    auto SkipNameCheck =
+        isa<ParmVarDecl>(ND1) || isa<ImplicitParamDecl>(ND1) ||
+        isa<ObjCTypeParamDecl>(ND1) || isa<TemplateTypeParmDecl>(ND1) ||
+        isa<TemplateTemplateParmDecl>(ND1) || isa<NonTypeTemplateParmDecl>(ND1);
+    if (!SkipNameCheck)
+      if (ND1->getQualifiedNameAsString() != ND2->getQualifiedNameAsString())
+        return false;
+  }
+

```